examples/window-app/nrfconnect/main/AppTask.cpp

```diff
@@ -0,0 +1,495 @@
+    }
+    else if (Instance().mFunctionTimerActive && Instance().mMode == OperatingMode::FactoryReset)
+    {
+        // Actually trigger Factory Reset
+        Instance().mMode = OperatingMode::Normal;
+        ConfigurationMgr().InitiateFactoryReset();
+    }
+}
+
+void AppTask::FunctionHandler(AppEvent * aEvent)
+{
+    if (!aEvent)
+        return;
+    if (aEvent->ButtonEvent.PinNo != FUNCTION_BUTTON)
+        return;
+
+    // To initiate factory reset: press the FUNCTION_BUTTON for FACTORY_RESET_TRIGGER_TIMEOUT + FACTORY_RESET_CANCEL_WINDOW_TIMEOUT
+    // All LEDs start blinking after FACTORY_RESET_TRIGGER_TIMEOUT to signal factory reset has been initiated.
+    // To cancel factory reset: release the FUNCTION_BUTTON once all LEDs start blinking within the
+    // FACTORY_RESET_CANCEL_WINDOW_TIMEOUT
+    if (aEvent->ButtonEvent.Action == BUTTON_PUSH_EVENT)
+    {
+        if (!Instance().mFunctionTimerActive && Instance().mMode == OperatingMode::Normal)
+        {
+            StartTimer(FACTORY_RESET_TRIGGER_TIMEOUT);
+        }
+    }
+    else
+    {
+        if (Instance().mFunctionTimerActive && Instance().mMode == OperatingMode::FactoryReset)
+        {
+            sUnusedLeds.Set(false);
+
+            UpdateStatusLED();
+            CancelTimer();
+
+            // Change the function to none selected since factory reset has been canceled.
+            Instance().mMode = OperatingMode::Normal;
+
+            LOG_INF("Factory Reset has been Canceled");
+        }
+        else if (Instance().mFunctionTimerActive)
+        {
+            CancelTimer();
+            Instance().mMode = OperatingMode::Normal;
+        }
+    }
+}
+
+void AppTask::StartBLEAdvertisementHandler(AppEvent *)
+{
+    if (Server::GetInstance().GetFabricTable().FabricCount() != 0)
+    {
+        LOG_INF("Matter service BLE advertising not started - device is already commissioned");
+        return;
+    }
+
+    if (ConnectivityMgr().IsBLEAdvertisingEnabled())
+    {
+        LOG_INF("BLE advertising is already enabled");
+        return;
+    }
+
+    if (Server::GetInstance().GetCommissioningWindowManager().OpenBasicCommissioningWindow() != CHIP_NO_ERROR)
+    {
+        LOG_ERR("OpenBasicCommissioningWindow() failed");
+    }
+}
+
+void AppTask::OpenHandler(AppEvent * aEvent)
+{
+    if (!aEvent)
+        return;
+    if (aEvent->ButtonEvent.PinNo != OPEN_BUTTON || Instance().mMode != OperatingMode::Normal)
+        return;
+
+    if (aEvent->ButtonEvent.Action == BUTTON_PUSH_EVENT)
+    {
+        Instance().mOpenButtonIsPressed = true;
+        if (Instance().mCloseButtonIsPressed)
+        {
+            Instance().ToggleMoveType();
+        }
+    }
+    else if (aEvent->ButtonEvent.Action == BUTTON_RELEASE_EVENT)
+    {
+        if (!Instance().mCloseButtonIsPressed)
+        {
+            if (!Instance().mMoveTypeRecentlyChanged)
+            {
+                WindowCovering::Instance().SetSingleStepTarget(OperationalState::MovingUpOrOpen);
```

Comment:
So this just sets the target position. But when the actual move happens? It's unclear to me how this is supposed to work.

Reply:
So it's either our UI (button 2/3) or the Matter controller which sets the target position. As a result there is a ZCL callback triggered and we react on that by starting the real/emulated movement of the shutter which in case of our app is just a timer. Every time the timer timeouts we consider the single step completed and set the current position accordingly (this triggers another callback in which we update the LED brightness - it indicates the current position of the cover). Now, from the UI we only support single steps, but from the Matter context we can get a target position which requires more single steps to complete, hence we restart the timer whenever the TargetCompleted() returns false. @Damian-Nordic I will appreciate if you can verify if this implementation does not cause any threads/async/Matter callback potential issues.

Reply:
Ok, I see it now. Thanks. I think the implementation looks OK now. I didn't look much into thread safety as that needs to be redesigned anyway. None of the examples does it right I think. Maybe switching to a single thread model entirely would be the easiest solution.